src/provider/provider_os_memory.c

```diff
@@ -0,0 +1,350 @@
+    [_UMF_OS_RESULT_ERROR_PURGE_LAZY_FAILED] = "lazy purging failed",
+    [_UMF_OS_RESULT_ERROR_PURGE_FORCE_FAILED] = "force purging failed",
+};
+
+static enum umf_result_t copy_nodemask(const unsigned long *nodemask,
+                                       unsigned long maxnode,
+                                       unsigned long **out_nodemask) {
+    if (out_nodemask == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (maxnode == 0 || nodemask == NULL) {
+        *out_nodemask = NULL;
+        return UMF_RESULT_SUCCESS;
+    }
+
+    size_t nodemask_bytes = (maxnode / 8) + ((maxnode % 8) ? 1 : 0);
+    size_t ul_size = sizeof(unsigned long);
+
+    // round to the next multiple of sizeof(unsigned long)
+    if (nodemask_bytes % ul_size != 0 || nodemask_bytes == 0) {
+        nodemask_bytes += ul_size - nodemask_bytes % ul_size;
+    }
+
+    *out_nodemask = (unsigned long *)calloc(1, nodemask_bytes);
+    if (!*out_nodemask) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    memcpy(*out_nodemask, nodemask, nodemask_bytes);
+
+    return UMF_RESULT_SUCCESS;
+}
+
+enum umf_result_t os_initialize(void *params, void **provider) {
+    if (provider == NULL || params == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    os_memory_provider_t *os_provider =
+        (os_memory_provider_t *)calloc(1, sizeof(os_memory_provider_t));
+    if (!os_provider) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    umf_os_memory_provider_params_t *os_params =
+        (umf_os_memory_provider_params_t *)params;
+    os_provider->params = *os_params;
+
+    enum umf_result_t ret = copy_nodemask(
+        os_params->nodemask, os_params->maxnode, &os_provider->params.nodemask);
+    if (ret != UMF_RESULT_SUCCESS) {
+        free(os_provider);
+        return ret;
+    }
+
+    *provider = os_provider;
+
+    return UMF_RESULT_SUCCESS;
+}
+
+void os_finalize(void *provider) {
+    if (provider == NULL) {
+        assert(0);
+        return;
+    }
+
+    os_memory_provider_t *os_provider = provider;
+    free(os_provider->params.nodemask);
+    free(os_provider);
+}
+
+static enum umf_result_t os_alloc(void *provider, size_t size, size_t alignment,
+                                  void **resultPtr) {
+    if (provider == NULL || resultPtr == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    os_memory_provider_t *os_provider = (os_memory_provider_t *)provider;
+    umf_os_memory_provider_params_t *os_params = &os_provider->params;
+
+    if (alignment && (alignment & (alignment - 1))) {
+        TLS_last_native_error = UMF_OS_RESULT_ERROR_WRONG_ALIGNMENT;
+        if (os_params->traces) {
+            fprintf(stderr, "wrong alignment (not a power of 2): %zu\n",
+                    alignment);
+        }
+        return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+    }
+
+    int flags = MAP_ANONYMOUS | os_params->visibility;
```

Comment:
why we always combine visibility with MAP_ANONYMUS? please add a comment for that

Reply:
`// MAP_ANONYMOUS - the mapping is not backed by any file`